sr_edc_ethercat_drivers/src/sr10.cpp

```diff
@@ -0,0 +1,573 @@
+  *  
+  *  @param this_buffer The data just being received by EtherCAT
+  *  @param prev_buffer The previous data received by EtherCAT
+  */
+bool SR10::unpackState(unsigned char *this_buffer, unsigned char *prev_buffer)
+{
+  ETHERCAT_DATA_STRUCTURE_0250_PALM_EDC_STATUS *status_data =
+          reinterpret_cast<ETHERCAT_DATA_STRUCTURE_0250_PALM_EDC_STATUS *>(this_buffer + command_size_);
+  ETHERCAT_CAN_BRIDGE_DATA *can_data = reinterpret_cast<ETHERCAT_CAN_BRIDGE_DATA *>(this_buffer + command_size_ +
+                                                                                    ETHERCAT_STATUS_DATA_SIZE);
+  // number of packets received since the start of the driver
+  static unsigned int num_rxed_packets = 0;
+
+  ++num_rxed_packets;
+
+  // Read IMU status and update internal structures
+  readImu(status_data);
+
+  // publishes the debug information (a slightly formatted version of the incoming ethercat packet):
+  if (debug_publisher->trylock())
+  {
+    debug_publisher->msg_.header.stamp = ros::Time::now();
+
+    debug_publisher->msg_.sensors.clear();
+    for (unsigned int i = 0; i < SENSORS_NUM_0220 + 1; ++i)
+    {
+      debug_publisher->msg_.sensors.push_back(status_data->sensors[i]);
+    }
+
+    debug_publisher->msg_.motor_data_type.data = static_cast<int> (status_data->motor_data_type);
+    debug_publisher->msg_.which_motors = status_data->which_motors;
+    debug_publisher->msg_.which_motor_data_arrived = status_data->which_motor_data_arrived;
+    debug_publisher->msg_.which_motor_data_had_errors = status_data->which_motor_data_had_errors;
+
+    debug_publisher->msg_.motor_data_packet_torque.clear();
+    debug_publisher->msg_.motor_data_packet_misc.clear();
+    for (unsigned int i = 0; i < 10; ++i)
+    {
+      debug_publisher->msg_.motor_data_packet_torque.push_back(status_data->motor_data_packet[i].torque);
+      debug_publisher->msg_.motor_data_packet_misc.push_back(status_data->motor_data_packet[i].misc);
+    }
+
+    debug_publisher->msg_.tactile_data_type = static_cast<unsigned int>(
+            static_cast<int32u>(status_data->tactile_data_type));
+    debug_publisher->msg_.tactile_data_valid = static_cast<unsigned int> (
+            static_cast<int16u> (status_data->tactile_data_valid));
+    debug_publisher->msg_.tactile.clear();
+    for (unsigned int i = 0; i < 5; ++i)
+    {
+      debug_publisher->msg_.tactile.push_back(
+              static_cast<unsigned int> (static_cast<int16u> (status_data->tactile[i].word[0])));
+    }
+
+    debug_publisher->msg_.idle_time_us = status_data->idle_time_us;
+
+    debug_publisher->unlockAndPublish();
+  }
+
+  if (status_data->EDC_command == EDC_COMMAND_INVALID)
+  {
+    // received empty message: the pic is not writing to its mailbox.
+    ++zero_buffer_read;
+    float percentage_packet_loss = 100.f * (static_cast<float>(zero_buffer_read) /
+                                            static_cast<float>(num_rxed_packets));
+
```

Comment:
Increment our new  `bad_frame_counter_` that will be published in the diagnostics (see other comment).

Reply:
Also here: Added here: https://github.com/shadow-robot/sr-ros-interface-ethercat/pull/265/commits/d65c6019a02b359ac64cffaae9b91fe8bea93719

Reply:
Removed for now, as these changes will be part of a separate PR: https://github.com/shadow-robot/sr-ros-interface-ethercat/pull/265/commits/8dffda4fd7ea2bdbc0e483260b80d1107dde9176